### PR TITLE
Error handling [proposal]

### PR DESCRIPTION
**Pull request by angetheo**
# Why
It seems semantically incorrect to treat everything between 407 and 500 as a fatal error.
Also, traversing the entire hash and find the status code inside a range to decide which error to raise is inefficient. Having a 1:1 mapping would allow us to retrieve the correct error code faster.

# What
- Add UnprocessableEntity error for `422`
- Add a method to retrieve a human-friendly message from the body of the error
- Map single status codes to errors without iterating over the whole hash

### CODE REVIEW

File: lib/routemaster/middleware/error_handling.rb [hunk 1]
```diff
@@ -5,23 +5,22 @@ module Routemaster
...
       }.freeze
 
       def on_complete(env)
-        ERRORS_MAPPING.each do |range, error_class|
-          if range.include?(env[:status])
-            raise error_class.new(env)
-          end
-        end
+        error_class = ERRORS_MAPPING[env[:status]]
+
```
Comment (danielcooper) [inline lib/routemaster/middleware/error_handling.rb]: Do we need a catch all for responses that aren't some form of OK and not in our error list?

**Review by danielcooper — APPROVED** [review]
➕ for getting rid of that scan

File: lib/routemaster/middleware/error_handling.rb (hunk 1 above)
Comment (angetheo) [reply lib/routemaster/middleware/error_handling.rb]: I thought about that as well, this list won't cover `407..421` and `423..499`. I've never seen such codes being used anywhere but rails might do something in the background for some particular requests that I might not know about. I guess it's better to play on the safe side and add a catch all for the remaining 4xx codes

File: spec/routemaster/integration/api_client_spec.rb [hunk 2]
```diff
@@ -151,11 +151,19 @@ def now
         expect { perform.(host + '/413') }.to raise_error(Routemaster::Errors::InvalidResource)
       end
 
+      it 'raises a FatalResource on uncaught error codes' do
+        expect { perform.(host + '/418') }.to raise_error(Routemaster::Errors::FatalResource)
```
Comment (gregbeech) [inline spec/routemaster/integration/api_client_spec.rb:155]: Errors::ImATeapot?

Comment (angetheo) [reply spec/routemaster/integration/api_client_spec.rb:155]: Ha, you found the easter egg! :egg:

**Pull request closed without merging**